syntax = "proto3";

package observer;

service Observer {
    rpc Publish(PublishRequest) returns (PublishResponse);
}

message PublishRequest {
    repeated Record records = 1;
}

message PublishResponse {
    bool successful = 1;
    optional string message = 2;
}

enum LogType {
    UNDECLARED_LOG_TYPE = 0;
    EVENT = 1;
    RUNTIME = 2;
    INPUT = 3;
    OUTPUT = 4;
    FEEDBACK = 5;
    METADATA = 6;
}

enum Tier {
    UNDECLARED_TIER = 0;
    SYSTEM = 1;
    SUBSYSTEM = 2;
    COMPONENT = 3;
    SUBCOMPONENT = 4;
}

message Record {
    Tier tier = 1;
    LogType log_type = 2;

    string parent_id = 3;
    string id = 4;

    // event
    optional string name = 5;
    optional string parameters = 6;
    optional string version = 7;
    optional string environment = 8;

    // runtime
    optional string start_time = 9;
    optional string end_time = 10;
    optional string error_type = 11;
    optional string error_content = 12;

    // input/output/feedback/metadata
    optional string field_name = 13;
    optional string field_value = 14;
}
